Add a lazy version of get() function (evaluate {default} lazily)

tyru opened the pull request:
`get()` is used also for initializing global variable when the variable does not exist.

```vim
let g:myplugin_globalvar = get(g:, 'myplugin_globalvar', 42)
```

But when I want to initialize the default value, the arguments for `get()` is evaluated before the initialization. 

```vim
" s:build_lazy() is evaluated even if g:myplugin_globalvar DOES EXIST.
let g:myplugin_globalvar = get(g:, 'myplugin_globalvar', s:build_lazy())
```

So I would like to propose `get_lazy()` which evaluates the 3rd argument lazily.

```vim
" s:build_lazy() is evaluated only when g:myplugin_globalvar does not exist.
let g:myplugin_globalvar = get_lazy(g:, 'myplugin_globalvar', {-> s:build_lazy()})
```

## Ref

https://github.com/vim/vim/pull/4695#issuecomment-513139727

tyru committed:
implement get_lazy()

tyru committed:
write help

tyru commented:
Hmm, some people told me that `get_lazy()` sounds a bit strange,
But I feel same myself... (but I can't find the good function name for it)
`get_lazy()` is `get()` lazily, but the main purpose of `get_lazy()` is to evaluate the argument lazily.
`build()` or `init()` is better? 🤔

tyru commented:
> `build()` or `init()` is better? 🤔

sorry it is too generic name.
`get_lazy()` is a bit strange but almost represents what it does 🤔

tyru commented:
Added ref link: https://github.com/vim/vim/pull/4695#issuecomment-513139727

ichizok commented:
How about `get_eval()` ?

tyru commented:
`get_eval()` sounds good 👍
How about this? >Bram, ALL

brammool commented:
> `get()` is used also for initializing global variable when the variable does not exist.
> 
> ```vim
> let g:myplugin_globalvar = get(g:, 'myplugin_globalvar', 42)
> ```
> 
> But when I want to initialize the default value, the arguments for `get()` is evaluated before the initialization. 
> 
> ```vim
> " s:build_lazy() is evaluated even if g:myplugin_globalvar DOES EXIST.
> let g:myplugin_globalvar = get(g:, 'myplugin_globalvar', s:build_lazy())
> ```
> 
> So I would like to propose `get_lazy()` which evaluates the 3rd argument lazily.
> 
> ```vim
> " s:build_lazy() is evaluated only when g:myplugin_globalvar does not exist.
> let g:myplugin_globalvar = get_lazy(g:, 'myplugin_globalvar', {-> s:build_lazy()})
> ```

Do we really need a new function for this?  We could just make get() do
this way.  I would be surprised scripts depend on evaluating a value
that isn't used.

-- 
"I know that there are people who don't love their fellow man,
and I hate those people!" - Tom Lehrer

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tyru commented:
> Do we really need a new function for this?  We could just make get() do
> this way.  I would be surprised scripts depend on evaluating a value
> that isn't used.

`get()`'s 3rd argument receives any type's value. it breaks backward compatibility.
`get()` is widely used for generic use. I think we cannot suppose it is not used.

brammool commented:
> > Do we really need a new function for this?  We could just make get() do
> > this way.  I would be surprised scripts depend on evaluating a value
> > that isn't used.
> 
> `get()`'s 3rd argument receives any type's value. it breaks backward
> compatibility.
> `get()` is widely used for generic use. I think we cannot suppose it
> is not used.

I thought it might be possible to postpone evaluating the argument until
it's needed.  Well, it is possible, but it is far from easy.  And if we
do this for internal functions, quickly the request comes to also
support it for user defined functions.

It's especially the error handling that becomes tricky.  And the
evaluation might throw an exception, which then needs to be handled.

get_lazy() does not sound so bad to me.
https://en.wikipedia.org/wiki/Lazy_evaluation

What was the objection for it?

-- 
Living on Earth includes an annual free trip around the Sun.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tyru commented:
> get_lazy() does not sound so bad to me.
> https://en.wikipedia.org/wiki/Lazy_evaluation
> 
> What was the objection for it?

I heard that from @ichizok, if I got it right from what he said:
I named `get_lazy()` because it means `get() for lazy`,
but `get_lazy()` appears `get lazy` (make (arguments) lazy).
But the argument of `get_lazy()` is lambda/funcref (which is already postponed).

The main purpose of this function is to "evaluate" the lazy (postponed) argument.
so I feel `get_eval()` is reasonable.

brammool commented:
> > get_lazy() does not sound so bad to me.
> > https://en.wikipedia.org/wiki/Lazy_evaluation
> > 
> > What was the objection for it?
> 
> I heard that from @ichizok, if I got it right from what he said:
> I named `get_lazy()` because it means `get() for lazy`,
> but `get_lazy()` appears `get lazy` (make (arguments) lazy).
> But the argument of `get_lazy()` is lambda/funcref (which is already postponed).
> 
> The main purpose of this function is to "evaluate" the lazy
> (postponed) argument.  so I feel `get_eval()` is reasonable.

For me the _eval part doesn't give a clue.  While lazy evaluation is a
common way to talk about evaluating something later.  So for me
get_lazy() has a nice hint.

I wonder if we can have a generic mechanism for lazy evaluation of
function arguments.  It can be very useful:

	func log(level, text)
	  if a:level > g:loglevel
	    call ch_log(text)
	  endif
	endfunc

If the "text" argument is marked as lazy it would be evaluated only when
used, which is more efficient.  The complexity of this is that "text"
needs to be evaluated in the scope of the caller:

	call log(INFO, "error in list: " .. string(thelist))

Passing a lambda works though and it's only a bit more text:

	call log(INFO, {-> "error in list: " .. string(thelist)})

Perhaps we can recommend that in the help.

-- 
ARTHUR: If you do not open these doors, we will take this castle by force ...
   [A bucket of slops land on ARTHUR.  He tries to retain his dignity.]
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tyru commented:
> I wonder if we can have a generic mechanism for lazy evaluation of
> function arguments.  It can be very useful:

Hmm, I don't think so... I think wrapping lambda like what you specified is enough.

```vim
call log(INFO, {-> "error in list: " .. string(thelist)})
```

The mechanism is called "Call by name".
https://en.wikipedia.org/wiki/Evaluation_strategy#Call_by_name

But I think introducing call by name evaluation strategy **to Vim script** causes many confusions...
In non-strict (lazy) evaluation languages, call by name or call by need (memoized version of call by name, evaluated only once) may help the optimization of speed.
But introducing call by name evaluation strategy to strict (non-lazy) languages like Vim script makes readers (of the code) confused because it is hard to understand the order of evaluation of call by name, and especially when lazy and non-lazy codes are mixed.

Lambda with no arguments like what you specified is a common technique to postpone evaluation in strict evaluation languages.
Vim script does not need call by name evaluation and it causes many confusions I think.

tyru commented:
> For me the _eval part doesn't give a clue.  While lazy evaluation is a
> common way to talk about evaluating something later.  So for me
> get_lazy() has a nice hint.

Okay, I think `get_lazy()` is also good name 👍

blueyed commented:
> I thought it might be possible to postpone evaluating the argument until it's needed. Well, it is possible, but it is far from easy.

That would indeed be nice.

As for the current approach: it works as `get_lazy(d, 'foo', 'GetDefaut()')`, correct?

I've thought it might make sense to support `v:val` and `v:key` there maybe, allowing for `get_lazy(d, 'foo', 'get(v:val, v:key)')`, e.g.:

> get_lazy(d, 'missing', 'GetDefault(v:val, v:key)')

`v:val` would be the first argument to `get()`, and `v:key` the 2nd.

While I am not sure how useful this is in general, it would allow for more flexible patterns.

brammool commented:
> > I thought it might be possible to postpone evaluating the argument until it's needed. Well, it is possible, but it is far from easy.
> 
> That would indeed be nice.
> 
> As for the current approach: it works as `get_lazy(d, 'foo',
> 'GetDefaut()')`, correct?

Without the (), the third argument is a function reference, and passing
the function name works.

> I've thought it might make sense to support `v:val` and `v:key` there
> maybe, allowing for `get_lazy(d, 'foo', 'get(v:val, v:key)')`, e.g.:
> 
> > get_lazy(d, 'missing', 'GetDefault(v:val, v:key)')
> 
> `v:val` would be the first argument to `get()`, and `v:key` the 2nd.

Hmm, in theory yes, but is this a practical example?  The reason we want
lazy evaluation is that it might do something expensive.  We might want
to pass even more arguments then.  The key is only needed if the
function would set the missing value as a side effect.

I think what you actually want is an auto-initializer, which is defined
on the dictionary instead of on the getter.  But that can be done by
defining a function on the dictionary that caches the result.  Does not
really require a new mechanism.

In case you do want to pass them, using a Lambda is quite easy:
    get_lazy(d, 'missing', {-> GetDefault(d, 'missing')})
    func GetDefault(dict, key)
      let val = HeavyComputation()
      let a:dict[a:key] = val
      return val
    endfunc

> While I am not sure how useful this is in general, it would allow for
> more flexible patterns.

I'm trying to avoid adding features that "someone may want to use
someday". 

-- 
"I've been teaching myself to play the piano for about 5 years and now write
most of my songs on it, mainly because I can never find any paper."
		Jeff Lynne, ELO's greatest hits

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tyru commented:
I agree with @brammool .

brammool commented:
OK, so please add a test.

tyru committed:
add get_lazy() tests

some weird behavior was found of get({func}, {what}) and
get_lazy({func}, {what}) ... see FIXME comments

tyru commented:
@brammool 
I added test cases, but found some bugs that `get()` and `get_lazy()` returns weird values for funcref (see `FIXME` comments).

```vim
" yes, it returned dict value
:echo type(get({-> 42}, 'dict')) == v:t_dict
" no output is produced?
:echo get({-> 42}, 'dict')
:echo string(get({-> 42}, 'dict'))
```

I'll fix this issue in another PR.

tyru committed:
doc: add get_lazy() example

tyru committed:
doc: add missing get_lazy({func}, {what})

blueyed commented:
>> As for the current approach: it works as `get_lazy(d, 'foo', 'GetDefaut()')`, correct?

> Without the (), the third argument is a function reference, and passing the function name works.

Do you mean `get_lazy(d, 'foo', GetDefaut)`?  That gives "undefined variable", e.g. with
```vim
let d = {}

fun! G(...)
  return a:000
endfun

echo get_lazy(d, 'foo', G)
```

>> `v:val` would be the first argument to `get()`, and `v:key` the 2nd.

> Hmm, in theory yes, but is this a practical example?

Not sure myself, it was just an idea.

>  The reason we want lazy evaluation is that it might do something expensive.
>  We might want to pass even more arguments then.

Do you mean "We might _not_ want to pass even more arguments then."?

> The key is only needed if the function would set the missing value as a side effect.

Or when looking it up from somewhere else, or returning something based on its name.

> In case you do want to pass them, using a Lambda is quite easy:
>    get_lazy(d, 'missing', {-> GetDefault(d, 'missing')})
>    func GetDefault(dict, key)

That however requires you to have a new lambda for every usage.

I've thought this might be useful for e.g. a generic helper function / expression, that could be re-used in different places (e.g. for looking up user settings).  Of course it could be wrapped (and likely is) in a function itself also.

I think I've thought about `v:key` / `v:val` in the first place, since it reminded me of map/filter in some way.

tyru committed:
fix test: 3rd arg is only for dict for funcref

tyru commented:
Added example and missing `get_lazy({func}, {what})` to help.

tyru committed:
fix code

tyru commented:
> Added example and missing `get_lazy({func}, {what})` to help.

Fixed (see below).

@brammool if you could include the fix in this PR, I will push this commit too.

```diff
diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
index 215bace32..894d573fd 100644
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -2394,10 +2394,10 @@ function({name} [, {arglist}] [, {dict}])
 garbagecollect([{atexit}])	none	free memory, breaking cyclic references
 get_lazy({list}, {idx} [, {deffunc}])	any	get item {idx} from {list} or result of {deffunc}
 get_lazy({dict}, {key} [, {deffunc}])	any	get item {key} from {dict} or result of {deffunc}
-get_lazy({func}, {what})		any	same as get({func}, {what})
+get_lazy({func}, {what} [, {deffunc}])		any	same as get({func}, {what})
 get({list}, {idx} [, {def}])	any	get item {idx} from {list} or {def}
 get({dict}, {key} [, {def}])	any	get item {key} from {dict} or {def}
-get({func}, {what})		any	get property of funcref/partial {func}
+get({func}, {what} [, {def})	any	get property of funcref/partial {func}
 getbufinfo([{expr}])		List	information about buffers
 getbufline({expr}, {lnum} [, {end}])
 				List	lines {lnum} to {end} of buffer {expr}
@@ -4727,7 +4727,7 @@ garbagecollect([{atexit}])				*garbagecollect()*
 get_lazy({list}, {idx} [, {defaultfunc}])
 get_lazy({blob}, {idx} [, {defaultfunc}])
 get_lazy({dict}, {key} [, {defaultfunc}])
-get_lazy({func}, {what})
+get_lazy({func}, {what} [, {defaultfunc}])
 		Same as |get()| but if {defaultfunc} was given and the item is not
 		available, call {defaultfunc} and return the result.
 >
@@ -4767,7 +4767,7 @@ get({dict}, {key} [, {default}])
 		Get item with key {key} from |Dictionary| {dict}.  When this
 		item is not available return {default}.  Return zero when
 		{default} is omitted.
-get({func}, {what})
+get({func}, {what} [, {default}])
 		Get an item with from Funcref {func}.  Possible values for
 		{what} are:
 			"name"	The function name
@@ -4775,6 +4775,9 @@ get({func}, {what})
 			"dict"	The dictionary
 			"args"	The list with arguments
 
+		Only when {what} == "dict" and item is not available, return
+		{default}. Return zero when {default} is omitted.
+
 							*getbufinfo()*
 getbufinfo([{expr}])
 getbufinfo([{dict}])
diff --git a/src/evalfunc.c b/src/evalfunc.c
index 899015843..4b252fe22 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -4237,6 +4237,7 @@ common_get(typval_T *argvars, typval_T *rettv, int is_lazy)
     dictitem_T	*di;
     dict_T	*d;
     typval_T	*tv = NULL;
+    int		what_dict = FALSE;
 
     if (argvars[0].v_type == VAR_BLOB)
     {
@@ -4310,7 +4311,11 @@ common_get(typval_T *argvars, typval_T *rettv, int is_lazy)
 		}
 	    }
 	    else if (STRCMP(what, "dict") == 0)
-		rettv_dict_set(rettv, pt->pt_dict);
+	    {
+		what_dict = TRUE;
+		if (pt->pt_dict != NULL)
+		    rettv_dict_set(tv, pt->pt_dict);
+	    }
 	    else if (STRCMP(what, "args") == 0)
 	    {
 		rettv->v_type = VAR_LIST;
@@ -4324,7 +4329,11 @@ common_get(typval_T *argvars, typval_T *rettv, int is_lazy)
 	    }
 	    else
 		semsg(_(e_invarg2), what);
-	    return;
+
+	    // When {what} == "dict", evaluate the third argument
+	    // if pt->pt_dict == NULL
+	    if (!what_dict)
+		return;
 	}
     }
     else
diff --git a/src/testdir/test_get_lazy.vim b/src/testdir/test_get_lazy.vim
index 51c192af8..881741d7f 100644
--- a/src/testdir/test_get_lazy.vim
+++ b/src/testdir/test_get_lazy.vim
@@ -31,44 +31,50 @@ func Test_get_lazy_blob()
   call assert_equal(999, result, 'result == 999')
 endfunc
 
-" get_lazy({lambda}, {what})
+" get_lazy({lambda}, {what} [, {defaultfunc}])
 func Test_get_lazy_lambda()
   let l:L = {-> 42}
   let l:Result = get_lazy(l:L, 'name')
   call assert_match('^<lambda>', l:Result, "l:Result =~ '^<lambda>'")
   let l:Result = get_lazy(l:L, 'func')
   call assert_equal(l:L, l:Result, "l:Result == l:L")
-  " FIXME: weird dict value was returned...
-  " let l:Result = get_lazy(l:L, 'dict', {-> {'lambda has': 'no dict'}})
-  " call assert_equal({}, l:Result, "l:Result == {'lambda has': 'no dict'}")
+  let l:Result = get_lazy(l:L, 'dict', {-> {'lambda has': 'no dict'}})
+  call assert_equal({'lambda has': 'no dict'}, l:Result,
+  \                 "l:Result == {'lambda has': 'no dict'}")
+  let l:Result = get_lazy(l:L, 'dict')
+  call assert_equal(0, l:Result, 'l:Result == 0')
   let l:Result = get_lazy(l:L, 'args')
   call assert_equal([], l:Result, "l:Result == []")
 endfunc
 
-" get_lazy({func}, {what})
+" get_lazy({func}, {what} [, {defaultfunc}])
 func Test_get_lazy_func()
   let l:F = function('tr')
   let l:Result = get_lazy(l:F, 'name')
   call assert_equal('tr', l:Result, "l:Result == 'tr'")
   let l:Result = get_lazy(l:F, 'func')
   call assert_equal(l:F, l:Result, "l:Result == l:F")
-  " FIXME: weird dict value was returned...
-  " let l:Result = get_lazy(l:F, 'dict', {-> {'func has': 'no dict'}})
-  " call assert_equal({}, l:Result, "l:Result == {'func has': 'no dict'}")
+  let l:Result = get_lazy(l:F, 'dict', {-> {'func has': 'no dict'}})
+  call assert_equal({'func has': 'no dict'}, l:Result,
+  \                 "l:Result == {'func has': 'no dict'}")
+  let l:Result = get_lazy(l:F, 'dict')
+  call assert_equal(0, l:Result, 'l:Result == 0')
   let l:Result = get_lazy(l:F, 'args')
   call assert_equal([], l:Result, "l:Result == []")
 endfunc
 
-" get_lazy({partial}, {what})
+" get_lazy({partial}, {what} [, {defaultfunc}])
 func Test_get_lazy_partial()
   let l:P = function('substitute', ['hello there', 'there'])
   let l:Result = get_lazy(l:P, 'name')
   call assert_equal('substitute', l:Result, "l:Result == 'substitute'")
   let l:Result = get_lazy(l:P, 'func')
   call assert_equal(function('substitute'), l:Result, "l:Result == function('substitute')")
-  " FIXME: weird dict value was returned...
-  " let l:Result = get_lazy(l:P, 'dict', {-> {'partial has': 'no dict'}})
-  " call assert_equal({}, l:Result, "l:Result == {'partial has': 'no dict'}")
+  let l:Result = get_lazy(l:P, 'dict', {-> {'partial has': 'no dict'}})
+  call assert_equal({'partial has': 'no dict'}, l:Result,
+  \                 "l:Result == {'partial has': 'no dict'}")
+  let l:Result = get_lazy(l:P, 'dict')
+  call assert_equal(0, l:Result, 'l:Result == 0')
   let l:Result = get_lazy(l:P, 'args')
   call assert_equal(['hello there', 'there'], l:Result, "l:Result == ['hello there', 'there']")
 endfunc
```

tyru commented:
Or you can retrieve whole diff including the fix from below URL.
(because now I pushed the branch to `get_lazy-and-fix-get-dict`)

Diff: https://github.com/vim/vim/compare/master...tyru:get_lazy-and-fix-get-dict.diff
PR style: https://github.com/vim/vim/compare/master...tyru:get_lazy-and-fix-get-dict

tyru commented:
@blueyed 
I said "I agree with @brammool .", but it was too few words and I remember one question I felt from your example.

I assume your first proposal is to support `v:key` and `v:val` in `get_lazy()`'s third argument.
https://github.com/vim/vim/pull/4696#issuecomment-514527716

But what value of `v:val`?
Maybe `v:key` is `'foo'` in the following example I think.
but there is no value for `v:val`.

```vim
let d = {}
" what values are returned?
echo get_lazy(d, 'foo', '[v:key, v:val]')
```

blueyed commented:
> But what value of `v:val`?

`v:val` would be the dict there:

```
let d = {}
" what values are returned?
echo get_lazy(d, 'foo', '[v:key, v:val]')

=> ['foo', {}]
```

tyru commented:
@blueyed 
Sorry for that I have misread your comment.
I thought you were saying that you agree with adding call by name evaluation.

@brammool 
If @blueyed's proposal is just only passing arguments, I agree with his proposal.
Because

* it is consistent with existing filter()/map() functions
* Using closure is good idea, but `function('GetDefaut')` is simpler than `{-> GetDefault(d, 'missing')}`
* And maybe little bit faster because it does not create a lambda for stub

tyru committed:
add more tests

Houl commented:
get_lazy() in Vim script

    func! GetLazy(dict, key, funcref)
        return has_key(a:dict, a:key) ? get(a:dict, a:key) : call(a:funcref, [])
    endfunc

tyru commented:
@Houl 
yeah, certainly it's easier to implement than C, but... often I want this in global variable initialization.
That's why I have added this generic function (and this PR does not make vim code bloated, just passing flag lazy or not).

However your example is only for dictionary value. it's needed to be more generic.
(I understand it works for almost case)

```vim
let s:NONE = []
function! s:get_lazy(container, key, ...)
    let l:Value = get(a:container, a:key, s:NONE)
    if l:Value isnot s:NONE
      return l:Value
    endif
    let l:F = get(a:000, 0, {-> 0})
    return type(l:F) is v:t_string ? s:call_string_expr(l:F, [a:container]) : l:F(a:key, a:container)
endfunction

" this is hack I wrote before in my library to evaluate v:key and v:val like built-in filter()/map().
" FIXME: in this case, v:val == a:container but v:key == 0 (expected: a:key).
function! s:call_string_expr(expr, args) abort
  return map([a:args[0]], a:expr)[0]
endfunction
```

@brammool also was wondering if this PR is necessary at first.
so I split #4733 bug fix to #4734 (maybe it's higher priority than this... ;)

tyru commented:
@brammool 

> If @blueyed's proposal is just only passing arguments, I agree with his proposal.

If this changes is OK, I'll fix the code and tests.

Houl commented:
@tyru 
I guess `container` can be a Dict or a List and if a key or index is missing you not only want to compute a default value, but also add it to the container?  If `container` is a List, then `key` is probably of no use (because if there is a gap, you need to fill that gap with other items)?  But if `container` is a Dict then your `s:F` needs both container and key as arguments?  For some reason, you omit `key` from the call if `s:F` is a string expression.

brammool commented:
Looking back at this now, I think we are adding something complicated to fix a simple problem.  It's more code, but doing what Houl mentioned is simpler.  And in most cases the default is just a value, in which case the overhead can be ignored.
So I tend to think we should not add get_lazy().

tyru commented:
Okay.
